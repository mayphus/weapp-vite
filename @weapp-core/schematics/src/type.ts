import type { App as _App, Component as _Component, Page as _Page, Sitemap as _Sitemap, Theme as _Theme } from './type.auto'

export type App = _App
export type Component = _Component
export type Page = _Page
export type Sitemap = _Sitemap
export type Theme = _Theme
