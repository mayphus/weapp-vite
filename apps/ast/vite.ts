import { build } from 'vite'

async function main() {
  await build({

    build: {
      rollupOptions: {
        input: {
          app: 'src/app.ts',
        },
        output: {
          chunkFileNames(chunkInfo) {
            return chunkInfo.name
          },
          entryFileNames(chunkInfo) {
            return `${chunkInfo.name}.js`
          },
          assetFileNames(chunkInfo) {
            return chunkInfo.names[0]
          },
        },

      },
      assetsDir: '',

    },
  })
}

main()
