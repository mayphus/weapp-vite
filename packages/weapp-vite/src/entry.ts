import type { Entry, WeappViteConfig } from './types'
import process from 'node:process'
import { defu } from '@weapp-core/shared'
import klaw from 'klaw'
import mm from 'micromatch'
import path from 'pathe'
import { defaultExcluded } from './defaults'
import { getWxmlEntry, searchAppEntry } from './utils'

export function createFilter(include: string[], exclude: string[], options?: mm.Options) {
  const opts = defu<mm.Options, mm.Options[]>(options, {
    ignore: exclude,
  })

  return function (id: unknown | string) {
    if (typeof id !== 'string') {
      return false
    }
    if (/\0/.test(id)) {
      return false
    }

    return mm.isMatch(id as string, include, opts)
  }
}

export async function getEntries(options: { root?: string, srcRoot?: string, outDir?: string, relative?: boolean, subPackage?: WeappViteConfig['subPackage'] }) {
  // build.outDir
  const { root = process.cwd(), outDir = 'dist', relative, srcRoot = '', subPackage } = options

  function formatPath(to: string) {
    if (relative) {
      return path.relative(root, to)
    }
    return path.normalize(to)
  }

  // 单独打包分包的场景
  if (subPackage) {
    const subPackageRoot = subPackage.root ?? ''
    const filter = createFilter(
      [path.join(root, srcRoot, subPackageRoot, '**/*')],
      [
        ...defaultExcluded,
      ],
      { cwd: root },
    )

    const pageEntries: Entry[] = []
    const componentEntries: Entry[] = []
    const subPackageEntries: Entry[] = []

    if (subPackage.entry) {
      subPackageEntries.push({
        deps: [],
        path: path.join(root, subPackageRoot, subPackage.entry),
        type: 'subPackageEntry',
      })
    }

    for await (
      const file of klaw(
        path.join(root, subPackageRoot),
        {
          filter,
        },
      )
    ) {
      if (file.stats.isFile()) {
        if (/\.wxml$/.test(file.path)) {
          const entry = getWxmlEntry(file.path, formatPath)
          if (entry) {
            //  && !walkPathsSet.has(file.path)
            // 防止重复，理论上不会
            // walkPathsSet.add(file.path)
            if (entry.type === 'component') {
              componentEntries.push(entry)
            }
            else if (entry.type === 'page') {
              pageEntries.push(entry)
            }
          }
        }
      }
    }
    return {
      pages: pageEntries,
      components: componentEntries,
      subPackageEntries,
    }
  }
  // 打包主包的场景
  else {
    const appEntry = searchAppEntry({
      root: path.join(root, srcRoot),
      formatPath,
    })

    // TODO exclude 需要和 output 配套
    // const walkPathsSet = new Set<string>()
    if (appEntry) {
      const subPackageDeps = appEntry.deps.filter(x => x.type === 'subPackage')
      const filter = createFilter(
        [path.join(root, srcRoot, '**/*')],
        [
          ...defaultExcluded,
          path.join(root, `${outDir}/**`),
          ...subPackageDeps.map((x) => {
            return path.join(root, `${x.root}/**`)
          }),
        ],
        { cwd: root },
      )

      const pageEntries: Entry[] = []
      const componentEntries: Entry[] = []

      for await (
        const file of klaw(
          path.join(root, srcRoot),
          {
            filter,
          },
        )
      ) {
        if (file.stats.isFile()) {
          if (/\.wxml$/.test(file.path)) {
            const entry = getWxmlEntry(file.path, formatPath)
            if (entry) {
              //  && !walkPathsSet.has(file.path)
              // 防止重复，理论上不会
              // walkPathsSet.add(file.path)
              if (entry.type === 'component') {
                componentEntries.push(entry)
              }
              else if (entry.type === 'page') {
                pageEntries.push(entry)
              }
            }
          }
        }
      }

      return {
        app: appEntry,
        pages: pageEntries,
        components: componentEntries,
        subPackages: subPackageDeps,
        // walkPathsSet,
      }
    }
  }
}
