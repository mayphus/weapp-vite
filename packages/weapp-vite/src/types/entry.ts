import type { App as AppJson, Component as ComponentJson, Page as PageJson, Sitemap, Theme } from '@weapp-core/schematics'

export interface BaseEntry {
  path: string
  jsonPath?: string
  json?: object
  type: 'app' | 'page' | 'component' | (string & {})
}

export type Entry = AppEntry | PageEntry | ComponentEntry

/**
 * 应用入口 js + json
 */
export interface AppEntry extends BaseEntry {
  type: 'app'
  themeJsonPath?: string
  themeJson?: Theme
  sitemapJsonPath?: string
  sitemapJson?: Sitemap
  json: AppJson
  jsonPath: string
}

/**
 * 页面入口 js + wxml
 */
export interface PageEntry extends BaseEntry {
  type: 'page'
  templatePath: string
  json?: PageJson
}

/**
 * 组件入口 js + wxml + json + json.component === true
 */
export interface ComponentEntry extends BaseEntry {
  type: 'component'
  templatePath: string
  json: ComponentJson
  jsonPath: string
}

export type EntryJsonFragment = Omit<BaseEntry, 'path' | 'type'>
