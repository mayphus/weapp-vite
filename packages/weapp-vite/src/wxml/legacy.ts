import type { Buffer } from 'node:buffer'
import type { ComponentsMap, ProcessWxmlOptions, WxmlDep } from '../types'
import { defu } from '@weapp-core/shared'
import { Parser } from 'htmlparser2'
import MagicString from 'magic-string'
import { isBuiltinComponent } from '../auto-import-components/builtin'
import { jsExtensions } from '../constants'
import { normalizeWxsFilename, transformWxsCode } from '../wxs'
import { srcImportTagsMap } from './shared'

// https://github.com/fb55/htmlparser2/issues/1541
// https://developers.weixin.qq.com/miniprogram/dev/framework/view/wxml/event.html
export function processWxml(wxml: string | Buffer, options?: ProcessWxmlOptions) {
  const opts = defu<Required<ProcessWxmlOptions>, ProcessWxmlOptions[]>(options, {
    excludeComponent: (tagName) => {
      return isBuiltinComponent(tagName)
    },
    platform: 'weapp',
    removeComment: true,
  })
  const ms = new MagicString(wxml.toString())
  const deps: WxmlDep[] = []
  let currentTagName = ''
  let importAttrs: undefined | string[]
  let attrs: Record<string, string> = {}
  const components: ComponentsMap = {}
  let tagStartIndex = 0
  // transformOn
  // https://github.com/vuejs/core/blob/76c43c6040518c93b41f60a28b224f967c007fdf/packages/compiler-core/src/transforms/vOn.ts

  const removeStartStack: number[] = []
  const removeEndStack: number[] = []
  const commentsPositions: { start: number, end: number }[] = []
  const parser = new Parser(
    {
      onopentagname(name) {
        currentTagName = name
        importAttrs = srcImportTagsMap[currentTagName]
        tagStartIndex = parser.startIndex
      },
      onattribute(name, value, quote) {
        attrs[name] = value
        if (importAttrs) {
          for (const attrName of importAttrs) {
            if (attrName === name) {
              deps.push({
                name,
                value,
                quote,
                tagName: currentTagName,
                start: parser.startIndex,
                end: parser.endIndex,
                attrs,
              })
              if (currentTagName === 'wxs' && name === 'src') {
                if (/\.wxs.[jt]s$/.test(value)) {
                  // 5 是 'src="'.length
                  // 1 是 '"'.length
                  ms.update(parser.startIndex + 5, parser.endIndex - 1, normalizeWxsFilename(value))
                }
              }
            }
          }
        }
        // 事件绑定
        if (name.startsWith('@')) {
          const start = parser.startIndex
          const end = parser.startIndex + name.length
          const { dir, mods } = name.split('.')
            .filter(x => x)
            .reduce<{ dir: string, mods: Record<string, boolean> }>((acc, cur, idx) => {
              if (idx === 0) {
                acc.dir = cur
              }
              else {
                acc.mods[cur] = true
              }
              return acc
            }, { dir: '', mods: {} })

          let rep: string
          if (mods.catch && mods.capture) {
            rep = `capture-catch:${dir.slice(1)}`
          }
          else if (mods.catch) {
            rep = `catch:${dir.slice(1)}`
          }
          else if (mods.mut) {
            rep = `mut-bind:${dir.slice(1)}`
          }
          else if (mods.capture) {
            rep = `capture-bind:${dir.slice(1)}`
          }
          else {
            rep = `bind:${dir.slice(1)}`
          }
          if (rep) {
            ms.update(start, end, rep)
          }
        }
        // 移除内联 wxs 的 lang
        if (currentTagName === 'wxs' && name === 'lang' && jsExtensions.includes(value)) {
          ms.update(parser.startIndex, parser.endIndex, '')
        }
      },
      onclosetag() {
        if (currentTagName && !opts.excludeComponent(currentTagName)) {
          if (Array.isArray(components[currentTagName])) {
            components[currentTagName].push({
              start: tagStartIndex,
              end: parser.endIndex + 1,
            })
          }
          else {
            components[currentTagName] = [{
              start: tagStartIndex,
              end: parser.endIndex + 1,
            }]
          }
        }

        currentTagName = ''
        attrs = {}
        importAttrs = undefined
        tagStartIndex = 0
      },
      ontext(data) {
        if (currentTagName === 'wxs' && jsExtensions.includes(attrs.lang)) {
          // data
          const res = transformWxsCode(data)
          if (res?.code) {
            ms.update(parser.startIndex, parser.endIndex, `\n${res.code}`)
          }
        }
      },
      // <!--  #ifdef  %PLATFORM% -->
      // 平台特有的组件
      // <!--  #endif -->
      oncomment(data) {
        let match = /#ifdef\s+(\w+)/.exec(data)
        if (match) {
          if (match[1] !== opts.platform) {
            removeStartStack.push(parser.startIndex)
          }
        }
        match = /#endif/.exec(data)
        if (match) {
          removeEndStack.push(parser.endIndex + 1)
        }
        commentsPositions.push({
          start: parser.startIndex,
          end: parser.endIndex + 1,
        })
      },
    },
    {
      lowerCaseTags: false,
      xmlMode: true,
    },
  )
  parser.write(
    ms.original,
  )
  parser.end()

  for (let i = 0; i < removeStartStack.length; i++) {
    const startIndex = removeStartStack[i]
    for (let j = i; j < removeEndStack.length; j++) {
      const endIndex = removeEndStack[j]
      if (endIndex > startIndex) {
        ms.remove(startIndex, endIndex)
        break
      }
    }
  }
  // remove comments
  if (opts.removeComment) {
    for (const { end, start } of commentsPositions) {
      ms.remove(start, end)
    }
  }

  return {
    components,
    deps,
    code: ms.toString(),
    removeStartStack,
    removeEndStack,
  }
}
