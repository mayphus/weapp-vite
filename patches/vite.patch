diff --git a/dist/node/chunks/dep-DDxXL6bt.js b/dist/node/chunks/dep-DDxXL6bt.js
index 1d5edd336a5216bd7b465392727d1434a97cfcb4..bd505effcd53ae1d74f3a70ac53621d14a24454c 100644
--- a/dist/node/chunks/dep-DDxXL6bt.js
+++ b/dist/node/chunks/dep-DDxXL6bt.js
@@ -48909,7 +48909,7 @@ function createCSSResolvers(config) {
   return {
     get css() {
       return cssResolve ??= createBackCompatIdResolver(config, {
-        extensions: [".css"],
+        extensions: [".css",".wxss"],
         mainFields: ["style"],
         conditions: ["style", DEV_PROD_CONDITION],
         tryIndex: false,
@@ -48919,7 +48919,7 @@ function createCSSResolvers(config) {
     get sass() {
       if (!sassResolve) {
         const resolver = createBackCompatIdResolver(config, {
-          extensions: [".scss", ".sass", ".css"],
+          extensions: [".scss", ".sass", ".css",".wxss"],
           mainFields: ["sass", "style"],
           conditions: ["sass", "style", DEV_PROD_CONDITION],
           tryIndex: true,
@@ -48942,7 +48942,7 @@ function createCSSResolvers(config) {
     },
     get less() {
       return lessResolve ??= createBackCompatIdResolver(config, {
-        extensions: [".less", ".css"],
+        extensions: [".less", ".css",".wxss"],
         mainFields: ["less", "style"],
         conditions: ["less", "style", DEV_PROD_CONDITION],
         tryIndex: false,
