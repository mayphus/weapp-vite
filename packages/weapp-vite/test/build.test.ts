// import { runProd } from '@/build'
import type { RollupOutput } from 'rollup'
import { sort } from 'fast-sort'
import path from 'pathe'
import { build } from 'vite'

import { getEntries } from './legacy/entry'

// import { createContext } from '@/context'

describe('build', () => {
  describe('rollup', () => {
    const mixjsDir = path.resolve(__dirname, './fixtures/mixjs')
    // function r(...args: string[]) {
    //   return path.resolve(mixjsDir, ...args)
    // }

    it('mixjs getEntries', async () => {
      const entries = await getEntries({
        root: mixjsDir,
        relative: true,
      })
      if (entries) {
        expect(entries.app).toMatchSnapshot('app')
        expect(sort(entries.pages).asc(x => x.path)).toMatchSnapshot('pages')
        expect(sort(entries.components).asc(x => x.path)).toMatchSnapshot('components')
        expect(entries.subPackageEntries).toMatchSnapshot('subPackageEntries')
        expect(entries.subPackages).toMatchSnapshot('subPackages')
      }
    })

    it('mixjs vite build', async () => {
      function relative(p: string) {
        return path.relative(mixjsDir, p)
      }
      const entries = await getEntries({
        root: mixjsDir,
      })
      if (!entries) {
        return
      }
      const paths = (entries.app ? [entries.app.path] : []).concat([...entries.pages, ...entries.components].map((x) => {
        return x.path
      }))
      const input = paths
        .reduce<Record<string, string>>((acc, cur) => {
          acc[relative(cur)] = cur
          return acc
        }, {})

      const res = await build({
        build: {
          rollupOptions: {
            input,
            output: {
              format: 'cjs',
              entryFileNames: (chunkInfo) => {
                return chunkInfo.name
              },
            },
          },
          assetsDir: '.',
          write: false,
          commonjsOptions: {
            include: undefined,
            transformMixedEsModules: true,
          },
          minify: false,
        },
        // plugins: [
        //   {
        //     name: 'inspect',
        //     enforce: 'pre',
        //     configResolved(_config) {
        //       // console.log(config)
        //     },
        //     resolveId(source) {
        //       console.log('resolveId', source)
        //     },
        //     load(id, options) {
        //       console.log('load', id, options)
        //     },
        //     generateBundle(_options, _bundle) {
        //       console.log('generateBundle', _bundle)
        //     },
        //   },
        // ],

      }) as RollupOutput

      for (const item of res.output) {
        // @ts-ignore
        expect(item.code).toMatchSnapshot(path.normalize(item.fileName))
      }
    })
  })
})
