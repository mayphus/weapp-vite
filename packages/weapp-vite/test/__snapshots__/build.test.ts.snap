// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`build > rollup > mixjs getEntries > app 1`] = `
{
  "deps": [
    {
      "path": "pages/index",
      "type": "page",
    },
    {
      "path": "pages/case0",
      "type": "page",
    },
    {
      "path": "/components/custom-bar2/index",
      "type": "component",
    },
    {
      "path": "tdesign-miniprogram/divider/divider",
      "type": "component",
    },
    {
      "entry": "entry.js",
      "pages": [
        "pages/cat",
        "pages/dog",
      ],
      "root": "packageA",
      "type": "subPackage",
    },
    {
      "name": "pack2",
      "pages": [
        "pages/apple",
        "pages/banana",
      ],
      "root": "packageB",
      "type": "subPackage",
    },
  ],
  "json": {
    "pages": [
      "pages/index",
      "pages/case0",
    ],
    "plugins": {
      "myPlugin": {
        "provider": "wxidxxxxxxxxxxxxxxxx",
        "version": "1.0.0",
      },
    },
    "subPackages": [
      {
        "entry": "entry.js",
        "pages": [
          "pages/cat",
          "pages/dog",
        ],
        "root": "packageA",
      },
      {
        "name": "pack2",
        "pages": [
          "pages/apple",
          "pages/banana",
        ],
        "root": "packageB",
      },
    ],
    "usingComponents": {
      "custom-bar2": "/components/custom-bar2/index",
      "t-divider": "tdesign-miniprogram/divider/divider",
    },
  },
  "jsonPath": "app.json",
  "path": "app.js",
  "type": "app",
}
`;

exports[`build > rollup > mixjs getEntries > components 1`] = `
[
  {
    "deps": [],
    "json": {
      "component": true,
    },
    "jsonPath": "components/custom-bar/index.json",
    "path": "components/custom-bar/index.ts",
    "type": "component",
  },
  {
    "deps": [],
    "json": {
      "component": true,
    },
    "jsonPath": "components/custom-bar1/index.json",
    "path": "components/custom-bar1/index.ts",
    "type": "component",
  },
  {
    "deps": [],
    "json": {
      "component": true,
    },
    "jsonPath": "components/custom-bar2/index.json",
    "path": "components/custom-bar2/index.ts",
    "type": "component",
  },
  {
    "deps": [],
    "json": {
      "component": true,
    },
    "jsonPath": "index2.json",
    "path": "index2.js",
    "type": "component",
  },
]
`;

exports[`build > rollup > mixjs getEntries > pages 1`] = `
[
  {
    "deps": [],
    "path": "index.js",
    "type": "page",
  },
  {
    "deps": [
      {
        "path": "../components/custom-bar/index",
        "type": "component",
      },
    ],
    "json": {
      "usingComponents": {
        "custom-bar": "../components/custom-bar/index",
      },
    },
    "jsonPath": "pages/case0.json",
    "path": "pages/case0.ts",
    "type": "page",
  },
  {
    "deps": [
      {
        "path": "/components/custom-bar1/index",
        "type": "component",
      },
      {
        "path": "plugin://myPlugin/hello-component",
        "type": "component",
      },
      {
        "path": "tdesign-miniprogram/button/button",
        "type": "component",
      },
    ],
    "json": {
      "usingComponents": {
        "custom-bar1": "/components/custom-bar1/index",
        "hello-component": "plugin://myPlugin/hello-component",
        "t-button": "tdesign-miniprogram/button/button",
      },
    },
    "jsonPath": "pages/index.json",
    "path": "pages/index.js",
    "type": "page",
  },
]
`;

exports[`build > rollup > mixjs getEntries > subPackageEntries 1`] = `undefined`;

exports[`build > rollup > mixjs getEntries > subPackages 1`] = `
[
  {
    "entry": "entry.js",
    "pages": [
      "pages/cat",
      "pages/dog",
    ],
    "root": "packageA",
    "type": "subPackage",
  },
  {
    "name": "pack2",
    "pages": [
      "pages/apple",
      "pages/banana",
    ],
    "root": "packageB",
    "type": "subPackage",
  },
]
`;

exports[`build > rollup > mixjs vite build > app.js 1`] = `
""use strict";
const cjs = require("./cjs-B3a6j7Rd.js");
const x = cjs.cjs;
console.log(x);
"
`;

exports[`build > rollup > mixjs vite build > cjs-B3a6j7Rd.js 1`] = `
""use strict";
const a = "1";
var cjs = {
  a,
  b: () => {
    return "b";
  }
};
exports.cjs = cjs;
"
`;

exports[`build > rollup > mixjs vite build > components/custom-bar/index.ts 1`] = `
""use strict";
Component({});
"
`;

exports[`build > rollup > mixjs vite build > components/custom-bar1/index.ts 1`] = `
""use strict";
Component({});
"
`;

exports[`build > rollup > mixjs vite build > components/custom-bar2/index.ts 1`] = `
""use strict";
Component({});
"
`;

exports[`build > rollup > mixjs vite build > esm-CR02SQDf.js 1`] = `
""use strict";
const c = "c";
exports.c = c;
"
`;

exports[`build > rollup > mixjs vite build > index.js 1`] = `
""use strict";
const cjs = require("./cjs-B3a6j7Rd.js");
const esm = require("./esm-CR02SQDf.js");
const a = cjs.cjs;
console.log(a, esm.c);
"
`;

exports[`build > rollup > mixjs vite build > index2.js 1`] = `
""use strict";
const cjs = require("./cjs-B3a6j7Rd.js");
const esm = require("./esm-CR02SQDf.js");
const a = cjs.cjs;
console.log(a, esm.c);
"
`;

exports[`build > rollup > mixjs vite build > pages/case0.ts 1`] = `
""use strict";
"
`;

exports[`build > rollup > mixjs vite build > pages/index.js 1`] = `
""use strict";
console.log("pages/index.wxml");
"
`;
